erpnext/regional/italy/utils.py

```diff
@@ -0,0 +1,175 @@
+		as_dict=1
+	)
+	invoice["company_address_data"]["country_code"] = frappe.db.get_value("Country",
+		invoice["company_address_data"]["country"], "code").upper()
+
+	invoice["company_contact_data"] = frappe.db.get_value("Company",
+		filters=invoice.company, 
+		fieldname=["phone_no", "email"], as_dict=1)
+
+	#Set invoice type
+	if invoice.is_return and invoice.return_against:
+		invoice["type_of_document"] = "TD04" #Credit Note (Nota di Credito)
+	else:
+		invoice["type_of_document"] = "TD01" #Sales Invoice (Fattura)
+	
+	#set customer information
+	invoice["customer_data"] = frappe.get_doc("Customer", invoice.customer)
+	invoice["customer_address_data"] = frappe.db.get_value("Address",
+		filters=invoice.customer_address,
+		fieldname=["address_line1", "city", "state", "country", "pincode"],
+		as_dict=1
+	)
+	invoice["customer_address_data"]["country_code"] = frappe.db.get_value("Country",
+		invoice["customer_address_data"]["country"], "code").upper()
+
+	if invoice.shipping_address_name:
+		invoice["shipping_address_data"] = frappe.db.get_value("Address",
+			filters=invoice.shipping_address_name,
+			fieldname=["address_line1", "city", "state", "country", "pincode"],
+			as_dict=1
+		)
+		invoice["shipping_address_data"]["country_code"] = frappe.db.get_value("Country",
+			invoice["shipping_address_data"]["country"], "code").upper()
+
+	if not invoice["vat_collectability"]:
+		invoice["vat_collectability"] = frappe.db.get_value("Company", invoice.company, "vat_collectability")
+
+	if invoice["customer_data"].is_public_administration:
+		invoice["transmission_format_code"] = "FPA12"
+	else:
+		invoice["transmission_format_code"] = "FPR12"
+	
+	#append items, and tax exemption reason.
+	items = frappe.get_all("Sales Invoice Item", filters={"parent":invoice.name}, fields=["*"], order_by="idx")
+	for item in items:
+		item_tax_rate = json.loads(item.item_tax_rate)
+		for account in item_tax_rate.keys():
+			item["tax_exemption_reason"] = frappe.db.get_value("Account", account, "tax_exemption_reason")
+
+	invoice["invoice_items"] = items
+
+	#tax rate wise grouping of tax amount and taxable amount.
+	invoice["tax_data"] = get_rate_wise_tax_data(invoice["invoice_items"])
+
+	invoice["payment_terms"] = frappe.get_all("Payment Schedule", filters={"parent": invoice.name}, fields=["*"])
+
+	return invoice
+
+def get_conditions(filters):
+	filters = json.loads(filters)
+
+	conditions = {"docstatus": 1}
+
+	if filters.get("company"): conditions["company"] = filters["company"]
+	if filters.get("customer"): conditions["customer"] =  filters["customer"]
+
+	if filters.get("from_date"): conditions["posting_date"] = (">=", filters["from_date"])
+	if filters.get("to_date"): conditions["posting_date"] = ("<=", filters["to_date"])
+	
+	if filters.get("from_date") and filters.get("to_date"):
+		conditions["posting_date"] = ("between", [filters.get("from_date"), filters.get("to_date")])
+	
+	return conditions
+
+def download_zip(files, output_filename):
```

Comment:
This function can be reused and moved to frappe

Reply:
https://github.com/frappe/frappe/pull/6853